test/integration/042_sources_test/test_sources.py

```diff
@@ -622,6 +674,107 @@ def test_run(self):
+        self.assertEqual(rowdict[0]['method'], 'run')
+        self.assertEqual(rowdict[0]['state'], 'running')
+        self.assertEqual(rowdict[0]['timeout'], None)
+
+        complete_ps_result = self.query('ps', completed=True, active=False).json()
+        result = self.assertIsResult(complete_ps_result)
+        self.assertEqual(len(result['rows']), 1)
+        self.assertEqual(len(result['rows'][0]), len(result['columns']))
+        rowdict = [{k: v for k, v in zip(result['columns'], row)} for row in result['rows']]
+        self.assertEqual(rowdict[0]['request_id'], 1)
+        self.assertEqual(rowdict[0]['method'], 'compile')
+        self.assertEqual(rowdict[0]['state'], 'finished')
+        self.assertEqual(rowdict[0]['timeout'], None)
+
+        all_ps_result = self.query('ps', completed=True, active=True).json()
+        result = self.assertIsResult(all_ps_result)
+        self.assertEqual(len(result['rows']), 2)
+        self.assertEqual(len(result['rows'][0]), len(result['columns']))
+        self.assertEqual(len(result['rows'][1]), len(result['columns']))
+        rowdict = [{k: v for k, v in zip(result['columns'], row)} for row in result['rows']]
+        rowdict.sort(key=lambda r: r['start'])
+        self.assertEqual(rowdict[0]['request_id'], 1)
+        self.assertEqual(rowdict[0]['method'], 'compile')
+        self.assertEqual(rowdict[0]['state'], 'finished')
+        self.assertEqual(rowdict[0]['timeout'], None)
+        self.assertEqual(rowdict[1]['request_id'], request_id)
+        self.assertEqual(rowdict[1]['method'], 'run')
+        self.assertEqual(rowdict[1]['state'], 'running')
+        self.assertEqual(rowdict[1]['timeout'], None)
+
+        self.kill_and_assert(pg_sleeper, sleep_task_id, request_id)
+
+    def kill_and_assert(self, pg_sleeper, task_id, request_id):
+        kill_result = self.query('kill', task_id=task_id).json()
+        kill_time = time.time()
+        result = self.assertIsResult(kill_result)
+        self.assertTrue(result['killed'])
+
+        sleeper_result = pg_sleeper.wait_result()
+        result_time = time.time()
+        error = self.assertIsErrorWithCode(sleeper_result, 10009, request_id)
+        self.assertEqual(error['message'], 'RPC process killed')
+        self.assertIn('data', error)
+        error_data = error['data']
+        self.assertEqual(error_data['signum'], 2)
+        self.assertEqual(error_data['message'], 'RPC process killed by signal 2')
+        self.assertIn('logs', error_data)
+        # it should take less than 5s to kill the process if things are working
+        # properly
+        self.assertLess(result_time, kill_time + 5)
+        return error_data
+
+    def _get_sleep_query(self):
+        request_id = 90890
+        pg_sleeper = self.background_query(
+            'run',
+            'select pg_sleep(15)',
+            _test_request_id=request_id,
+            name='sleeper',
+        )
+
+        for _ in range(20):
+            time.sleep(0.2)
```

Comment:
There is a deadlock in tests, maybe sometimes. I think adding this sleep to the test loop fixed it, which does not make me feel good. But I tested it by running the tests over and over in a loop and they don't fail, which does make me feel a little better at least.